[DVCSMP-3413] Implemented Lock Code Scheduling functionality.

vaish-rahul opened the pull request:
info so that the Zigbee Locks get adapted to Smart Locks Version 2
features.

vaish-rahul commented:
@greens, @mckeed Please review the changes.

vaish-rahul commented:
This PR is dependent on following PR: https://github.com/SmartThingsCommunity/SmartThingsCapabilities/pull/248

greens commented on devicetypes/smartthings/zigbee-lock.src/zigbee-lock.groovy:
```diff
@@ -575,10 +578,17 @@ private def parseCommandResponse(String description) {
 				break
 			case 2:
 				responseMap.value = "unlocked"
+				if (responseMap.data) {
```
Why not just one line here?

`responseMap.data = [notify : true]`

Is there a subtlety I'm missing?

vaish-rahul replied:
It's possible that responseMap.data already has some values in it by the time it reaches this line. For e.g. at line number 560 the responseMap.data can get assigned with a map and if that happens then doing `responseMap.data = [notify : true]` will overwrite the existing map with new one. To avoid this problem `responseMap.data.notify = true` is done which will just add another key-value pair to the existing map.

greens commented on devicetypes/smartthings/zigbee-lock.src/zigbee-lock.groovy:
```diff
@@ -598,7 +608,13 @@ private def parseCommandResponse(String description) {
 			case 9: // unlocked using the key
 			case 14: // unlocked using the Thumbturn
 				responseMap.value = "unlocked"
+				if (responseMap.data) {
```
same question as above

vaish-rahul replied:
same answer as above.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1874,568 @@ def deleteCode(codeID) {
+ *
+ * @return The command(s) for deleting week day schedule
+ */
+private def getWeekDayDeleteScheduleCmd(schSettings) {
```
Does this still need to be defined?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,374 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+				def stopDay = (cmd.stopDay < 10) ? "0${cmd.stopDay}" : "${cmd.stopDay}"
+				def stopHour = (cmd.stopHour < 10) ? "0${cmd.stopHour}" : "${cmd.stopHour}"
+				def stopMinute = (cmd.stopMinute < 10) ? "0${cmd.stopMinute}" : "${cmd.stopMinute}"
+				state["setYearDayStartDate$codeID"] = "20${startYear}-${startMonth}-${startDay}"
```
You could save a lot of lines using:

`sprintf("%02d", variableName)`

greens replied:
or in this case specifically:

`sprintf("20%02d-%02d-%02d", startYear, startMonth, startDay)`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1874,568 @@ def deleteCode(codeID) {
+					codesToScanSchedules << i
+				}
+			}
+			state.codesToScanSchedules = codesToScanSchedules
```
In what cases would this not be equivalent to `lockCodes.keySet()`?

vaish-rahul replied:
We scan only the first 8 slots when the lock is paired freshly. Among the first 8 slots it is possible that only few slots have lock code for e.g. only slot 1, 3, 4, 8. In this case the `codesToScanSchedules` list will contain 1, 3 ,4, 8 and for these slots schedules will be scanned. Meanwhile a user can create lock code in other available slots, either from smartapp or directly on the lock. In this case the new slots will get added to `lockCodes` map but we shouldn't be scanning those slots.

greens replied:
how about 

`codesToScanSchedules = keySet.findall {it <= MAX_CODES_TO_SCAN}`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1874,568 @@ def deleteCode(codeID) {
+	if (codesToScanSchedules && codesToScanSchedules.size() > 0) {
+		return true
+	}
+	return false
```
`return state.codesToScanSchedules && state.codesToScanSchedules.size() > 0` ?

or

`return states.codesToScanSchedules?.size() > 0`

vaish-rahul commented:
@greens I have made changes according to your review comments. Please check.

vaish-rahul commented:
@greens, @jodyalbritton Any other review comments on this PR?

greens commented:
Lots to get through. Still working on it in-between other things.

mckeed commented on devicetypes/smartthings/zigbee-lock.src/zigbee-lock.groovy:
```diff
@@ -1176,3 +1197,72 @@ private boolean isMasterCode(codeID) {
+		codeID = codeID.toString()
+		def lockCodesData = loadlockCodesData(codeID)
+		lockCodesData[codeID].lastUsage = userInfo.lastUsage
+		lockCodesData[codeID].userInfo.phoneNum = userInfo.phoneNum ?: ""
```
What is phoneNum used for? This doesn't seem like the right place to be storing phone numbers.

vaish-rahul replied:
In DTH we store code name for e.g. "Plumber" etc., in lockCodes attribute, which is actually the guest name. This was implemented in Smart Locks V1. The reason was that if user uninstalls the smartapp and installs it again then lockCodes data shouldn't get erased and hence it was stored in DTH.

In Smart Locks V2 this is extended to storing guest's phone number also. If we shift the phone number to SmartApp then the solution will break if user uninstalls and installs the SmartApp. In this case the code name will be picked up from DTH but phone number will be missing.

vaish-rahul committed:
[DVCSMP-3413] Implemented Lock Codes Scheduling functionality in
Z-Wave Locks DTH for supporting Smart Locks Version 2 features. Z-Wave
Locks will support Always working, Daily Repeating, Week Day and Year
Day scheduled lock codes.

Adapted Zigbee Locks DTH also for supporting Smart Locks Version 2
features. Zigbee Locks will support only Always working lock codes.

Made changes according to the review comments.

Changed lastUsage to lastUpdated as per the review comments.

greens commented on devicetypes/smartthings/zigbee-lock.src/zigbee-lock.groovy:
```diff
@@ -1176,3 +1197,72 @@ private boolean isMasterCode(codeID) {
+	if (codeID) {
+		codeID = codeID.toString()
+		def lockCodesData = loadlockCodesData(codeID)
+		lockCodesData[codeID].lastUpdated = userInfo.lastUpdated
```
Is the `lastUpdated` field being used for anything?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -24,7 +24,7 @@ metadata {
 		capability "Battery"
 		capability "Health Check"
 		capability "Configuration"
```
Once we decide the name of the capability, it should be added here, right?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -169,11 +189,37 @@ def doConfigure() {
+	cmds << setClock()
+	cmds << configureScheduling()
+	// delayBetween does the flattening
+	cmds = delayBetween(cmds, 200)
```
Just pointing out that we've had issues elsewhere with short delays between lock commands. I'd prefer that this value at least remained `4200`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+def zwaveEvent(physicalgraph.zwave.commands.scheduleentrylockv3.ScheduleEntryTypeSupportedReport cmd) {
+	log.trace "[DTH] Executing 'zwaveEvent(physicalgraph.zwave.commands.scheduleentrylockv3.ScheduleEntryTypeSupportedReport cmd)' with cmd = $cmd"
+	def result = []
+	def numberOfSlotsDailyRepeating = cmd.numberOfSlotsDailyRepeating
```
Can we do:

`def numberOfSlotsDailyRepeating = cmd.numberOfSlotsDailyRepeating ?: 0`

here?

greens replied:
Or even just put that statement in the `descriptionText` directly?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+			def changeType = getScheduleChangeType(codeID, SCHEDULE_TYPE_DAILY_REPEATING)
+			
+			if (savedStartTime && savedEndTime && savedWeekDayBitmask) {
+				def startTime = getWeekDayHoursAndMinutes(savedStartTime)
```
Every time you call `getWeekDayHoursAndMinutes` you're immediately breaking it up into these separate variables. You should either change the signature to just return a list and then do:

`def (startHour, startMin) = getWeekDayHoursAndMinutes(savedStartTime)`

or just reference startTime's individual attributes every time.

mckeed replied:
Unfortunately the destructuring assignment syntax `def (startHour, startMin)` is broken by the sandbox parser, or at least was last time I used it. I agree just using startTime.hour is stylistically better than startHour, though.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+				def endHour = endTime.hour
+				def endMin = endTime.minute
+		
+				def durMap = getDurationInHoursAndMinutes(startHour, startMin, endHour, endMin)
```
same rationale for `getDurationInHoursAndMinutes`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+		def codeName = getCodeName(lockCodes, codeID)
+		
+		def daysOfWeek = state["setWeekDayWeekDays$codeID"]
+		if (daysOfWeek && daysOfWeek.size() > 0) {
```
`daysOfWeek?.size() > 0`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+				def endHour = endTime.hour
+				def endMin = endTime.minute
+				
+				if (cmd.startHour == startHour && cmd.startMinute == startMin && cmd.stopHour == endHour && cmd.stopMinute == endMin) {
```
As a general comment, can this be commented a bit more? It's not entirely clear at a glance what is happening in this code.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+			}
+		} else {
+			def scheduleSlotId = cmd.scheduleSlotId
+			scheduleSlotId = scheduleSlotId - 1
```
just do this in one line

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1035,6 +1163,366 @@ def zwaveEvent(physicalgraph.zwave.commands.applicationstatusv1.ApplicationRejec
+			def changeType = getScheduleChangeType(codeID, SCHEDULE_TYPE_YEAR_DAY)
+			
+			if (startDate && startTime && endDate && endTime) {
+				def startDateTime = getYearDayTimeDetails(startDate, startTime)
```
same rationale as `getDurationInHoursAndMinutes` changes.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+	log.trace "[DTH] Executing 'startScanningSchedule()' by ${device.displayName}"
+	def codesToScanSchedules = state.codesToScanSchedules
+	def codeID = -1
+	if (codesToScanSchedules && codesToScanSchedules.size() > 0) {
```
`codesToScanSchedules?.size() > 0`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+		def lockCodes = loadLockCodes()
+		if (lockCodes && lockCodes.size() > 0) {
+			def keySet = lockCodes.keySet()
+			codesToScanSchedules = []
```
In the creation/assignment above, you could just do

`def codesToScanSchedules = state.codesToScanSchedules ?: []`

and avoid this line.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+	if (isKwiksetLock()) {
+		// For Kwikset locks we can find out the user type and then issue the appropriate schedule get command
+		log.trace "[DTH] scanSchedule() -  Calling configurationGet() for codeID := $codeID"
+		sendHubCommand(new physicalgraph.device.HubAction(secure(zwave.configurationV2.configurationGet(parameterNumber: codeID))))
```
I added some code recently so that you shouldn't have to cast Zwave commands to HubActions anymore for `sendHubCommand`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+ * @returns cmds: The commands fired for changing the user type
+ */
+def changeUserTypeToOwner(codeID) {
+	if (isKwiksetLock()) {
```
`if (isKwiksetLock() && codeID)`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+		def startHour = startTime.hour
+		def startMin = startTime.minute
+		
+		def endTime = getWeekDayHoursAndMinutes(schSettings.endTime)
```
not used in this scope

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+			if (isBitSet(weekDayBitMask, i)) {
+				daysOfWeek << i
+				setCmds << secure(zwave.scheduleEntryLockV3.scheduleEntryLockWeekDaySet(startHour: startHour, startMinute: startMin,
+					setAction: 1, scheduleSlotId: (i + 1), stopHour: endHour, stopMinute: endMin, userIdentifier: codeID, dayOfWeek: i))
```
what's the relation between the `scheduleSlotId` and `dayOfWeek`?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+		def startDay = startDateTime.day
+		def startMonth = startDateTime.month
+		def startYear = startDateTime.year.toString()
+		startYear = startYear.substring(2, startYear.length())
```
If you're just trying to get the last 2 digits of the year and `startDateTime` is already an Integer:

`def startYear = startDateTime.year % 100`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+		stopYear = stopYear.substring(2, stopYear.length())
+		stopYear = Integer.parseInt(stopYear, 10)
+		
+		cmds << secure(zwave.scheduleEntryLockV3.scheduleEntryLockYearDaySet(scheduleSlotId: 1, setAction: 1, startDay: startDay, startHour: startHour, startMinute: startMinute,
```
As a general comment, you have a bunch of these functions where you break out all your arguments into separate assignments and then just use them once as arguments to a function call.

I'd recommend just referencing them directly instead of having the extra assignment.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+		setCmds << secure(zwave.scheduleEntryLockV3.scheduleEntryLockWeekDaySet(setAction: 0, scheduleSlotId: (i + 1), userIdentifier: codeID))
+		getCmds << requestWeekDaySchedule((i + 1), codeID)
+	}
+	if (setCmds && getCmds) {
```
Is there any way both of these won't be populated?

greens replied:
Also doesn't the `requestAllWeekDaySchedule` function below generate `getCmds` exactly?

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+ */
+def isBitSet(weekDayBitmask, bitNum) {
+	def mask = 1 << bitNum
+	if ((weekDayBitmask & mask) != 0) {
```
`return (weekDayBitmask & (1 << num)) != 0`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+
+	long difference = endDate.getTime() - startDate.getTime();
+	difference = difference / 1000
+	def durSeconds = difference % 60
```
Check my math, but I'm pretty sure `durSeconds` is always 0.

greens replied:
I think the better approach here, since you're always dealing in full minutes and hours, is to compute using minutes:

`def difInMinutes = ((endHour * 60) + endMin) - ((startHour * 60) + startMin)`
`def durMin = difInMinutes % 60`
`def durHour = (difInMinutes / 60) % 24)`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+def clearUserInfo(codeID) {
+	log.trace "[DTH] Executing 'clearUserInfo()' by ${device.displayName} with codeID := $codeID"
+	def lockCodesData = parseJson(device.currentValue("lockCodesData") ?: "{}") ?: [:]
+	codeID = codeID.toString()
```
as a general style comment, very simple statements like this don't need to be broken out into one line

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1365,6 +1866,579 @@ def deleteCode(codeID) {
+ */
+def clearUserInfo(codeID) {
+	log.trace "[DTH] Executing 'clearUserInfo()' by ${device.displayName} with codeID := $codeID"
+	def lockCodesData = parseJson(device.currentValue("lockCodesData") ?: "{}") ?: [:]
```
near-duplication of `loadLockCodesData()`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1526,6 +2631,54 @@ private def codeSetEvent(lockCodes, codeID, codeName) {
+	def lockCodesData = loadlockCodesData(codeID)
+	lockCodesData[codeID].lastUpdated = new Date().getTime()
+	lockCodesData[codeID].scheduleInfo = [:]
+	lockCodesData[codeID].scheduleInfo.scheduleData = [:]
```
Don't bother clearing this if you're just going to re-assign.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1588,7 +2741,73 @@ private def getChangeType(lockCodes, codeID) {
+ */
+private def getScheduleChangeType(codeID, scheduleType) {
+	def changeType = "set"
+	def lockCodesData = device.currentValue("lockCodesData")
```
Some of these lines can be replaced with a `loadLockCodesData()` call.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1588,7 +2741,73 @@ private def getChangeType(lockCodes, codeID) {
+	if (lockCodesData) {
+		lockCodesData = parseJson(lockCodesData)
+		def data = lockCodesData[codeID.toString()]
+		if (data) {
```
this should really be

`if (data?.scheduleInfo?.scheduleData)`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1588,7 +2741,73 @@ private def getChangeType(lockCodes, codeID) {
+ */
+private def isScheduleScanningInProgress(codeID) {
+	def codesToScanSchedules = state.codesToScanSchedules
+	if (codesToScanSchedules && codeID.toInteger() in codesToScanSchedules) {
```
`return (state.codesToScanSchedules && (codeID.toInteger() in state.codesToScanSchedules))`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1588,7 +2741,73 @@ private def getChangeType(lockCodes, codeID) {
+ */
+private def getScheduleType(codeID) {
+	def scheduleType = "always"
+	def lockCodesData = device.currentValue("lockCodesData")
```
`loadLockCodesData()`

greens replied:
`scheduleType = lockCodesData[codeID.toString()]?.scheduleInfo?.scheduleType ?: "always"`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1602,6 +2821,40 @@ private def getStatusForDescription(changeType) {
+ *
+ * @return The map with hour and minute
+ */
+private def getWeekDayHoursAndMinutes(time) {
```
if the time string is formatted as "HH:mm", it's much simpler to just do:

`time = time.split(':')`
`return [hour: time[0] as Integer , minute: time[1] as Integer]`

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1602,6 +2821,40 @@ private def getStatusForDescription(changeType) {
+ * @return The map with year, month, day, hour and minute
+ */
+private def getYearDayTimeDetails(date, time) {
+	def sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm")
```
I might recommend doing the same splitting as above here as well on the date (using a hyphen), since you're just putting them into the `Calendar` and the immediately pulling them back out.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1695,3 +3006,53 @@ def readCodeSlotId(physicalgraph.zwave.commands.alarmv2.AlarmReport cmd) {
+			def minutesDst = dstOffset  / 60
+			log.debug "Setting offset on lock hourTzo: ${hours} minuteOffsetDst: ${minutesDst} minuteTzo: ${minutes} signOffsetDst: ${dstFlag} signTzo: ${timeZoneFlag}"
+			cmds << secure(zwave.scheduleEntryLockV3.scheduleEntryLockTimeOffsetSet(hourTzo: hours, minuteOffsetDst: minutesDst, minuteTzo: minutes, signOffsetDst: dstFlag, signTzo: timeZoneFlag))
+			}
```
fix indentation

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1695,3 +3006,53 @@ def readCodeSlotId(physicalgraph.zwave.commands.alarmv2.AlarmReport cmd) {
+
+		if(!location.timeZone.useDaylightTime() || !location.timeZone.inDaylightTime(new Date())) {
+			log.debug "Setting offset on lock hourTzo: ${hours} minuteOffsetDst: 0 minuteTzo: ${minutes} signOffsetDst: 0 signTzo: ${timeZoneFlag}"
+			cmds << secure(zwave.scheduleEntryLockV3.scheduleEntryLockTimeOffsetSet(hourTzo:hours, minuteOffsetDst:0, minuteTzo:minutes, signOffsetDst:0, signTzo: timeZoneFlag))
```
default values (`minuteOffsetDst` and `signOffsetDst`) can be left out here.

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1695,3 +3006,53 @@ def readCodeSlotId(physicalgraph.zwave.commands.alarmv2.AlarmReport cmd) {
+		int year = now.get(Calendar.YEAR)
+
+		log.debug "Setting time on lock - $year-$month-$day $hour:$min:$sec (YYYY-MM-DD hh:mm:ss)"
+		cmds << secure(zwave.timeParametersV1.timeParametersSet(hourUtc: hour, minuteUtc: min, secondUtc: sec, day: day , month: month, year: year))
```
just reference the values of the calendar object directly inline

greens commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -1612,6 +2865,64 @@ def clearStateForSlot(codeID) {
+ * @param codeID: The code slot number
+ */
+def clearDailyRepeatingScheduleDataFromState(codeID) {
+	state["setDailyRepeatingStartTime$codeID"] = null
```
A couple of things about this style of state storage:

Why not just store `state["code$codeID"]` as a dict?

I think that would make adding/removing it a lot simpler.

Also I'd check to see that both of these steps are necessary. It would be my instinct that just calling `state.remove` should get rid of the data stored in the field without you having to specifically null it out.

greens reviewed (changes requested):
I think you've got enough here for a first pass.

Keeping in mind that this is largely just a style review.

Do you think you could go through for me the flow from adding a device to setting a code through some app, and then deleting that code?

mckeed commented on devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy:
```diff
@@ -386,15 +487,15 @@ private def handleAccessAlarmReport(cmd) {
 		case 2: // Manually unlocked
 			map.descriptionText = "Unlocked manually"
-			map.data = [ method: "manual" ]
+			map.data = [ method: "manual", notify: true, notificationText: "$deviceName at ${location.name} was unlocked manually" ]
```
What is the purpose of `notify` in the event data? Can't that be handled by the SmartApp?
